### PR TITLE
improve smooth module scrolling logic

### LINKED ISSUE #12686: unwanted scroll of module making it unusable
On an HD screen doing a demonstration tonight I stumbled on a bug probably due to the new scrolling of module introduced recently.

The scrolling was making the module unusable, it was on an HD screen, and scrolling was activated when I was flying over the blend section. I have been able to reproduce this consistently following the procedure below:

To reproduce:
- On the RVB curve module
- Select drawn & parametric blend
- Scroll down the panel to be able to show all controls of the blend section and make sure the header of the module is not visible anymore.
- Fly over the parametric blend canal (g R V B ..)
- See the module scroll down to show the header on top... making the blend section outside of the panel (I was on a small HD screen).

I think this is due to #12637, @dterrahe sounds like this is for you :)

### PR DESCRIPTION
**Pull request by dterrahe**
I believe this fixes #12686 but please test all scenarios carefully because there might still be corner cases I haven't thought of.

I haven't found a way to prevent the resize and repositioning triggered by hovering over the notebook tabs or manipulating the channel controls. So whenever a resize happens, there might still be a move. What this _does_ try to do is avoid scrolling any part of the active module (further) out of view. It will still try to show as much as possible of the module if anything was hidden that there is (now) space for or, in case "scroll to top" is on, move the module to the top again if that doesn't obscure anything currently visible. Hopefully some of the "unnecessary" scrolling is not too annoying and maybe even helpful.

For me, switching off "scroll to top" gives a better experience overall than having it on, as these changes make it (imvho of course) no longer that useful.

### CODE REVIEW

**Comment by da-phil** [comment]
I can confirm that it fixes the bug for me.

**Comment by da-phil** [comment]
By the way, is there a way how to disable the opening/collapsing animation of modules? For me this is an annoying gadget.

**Comment by dterrahe** [comment]
Yes; set duration to 0ms in preferences

**Comment by da-phil** [comment]
> Yes; set duration to 0ms in preferences

Ups, missed this one, thanks

**Comment by dterrahe** [comment]
> Ups, missed this one, thanks

It will be in the release notes.

**Review by TurboGit — APPROVED** [review]
Works for me, tested many scenario, all good. Thanks!

File: src/dtgtk/expander.c [hunk 1]
```diff
@@ -121,21 +121,24 @@ static gboolean _expander_scroll(GtkWidget *widget, GdkFrameClock *frame_clock,
...
   }
 
   float prop = 1.0f;
-  gboolean scroll_to_top = dt_conf_get_bool("darkroom/ui/scroll_to_module");
-  if((allocation.y + allocation.height - value > available.height && value < allocation.y) ||
-     (scroll_to_top && allocation.y != value))
+  const gboolean scroll_to_top = dt_conf_get_bool("darkroom/ui/scroll_to_module");
+  const int spare = available.height - allocation.height;
+  const int from_top = allocation.y - value;
+  const int move = MAX(from_top - scroll_to_top ? 0 : MAX(0, MIN(from_top, spare)),
```
[the remark this answers is not in the document]
Comment (TurboGit) [reply src/dtgtk/expander.c:133]: > Does it still work for you? Should I submit a new fix? After more testing? I'm just quite confused.

No it doesn't work for me either. I'm not sure I have tested this scenario though. I was looking at the fact that it should not scroll when I use the blend controls, but I've probably done that on a small module that does not extends much on the bottom.

[the remark this answers is not in the document]
Comment (dterrahe) [reply src/dtgtk/expander.c:133]: But does it "scroll to top" when expanding a module? (if you have that option selected in prefs) For me it basically doesn't scroll ever...

It's not that it surprises me that the code as is doesn't work. It is that I recall testing it extensively and (obviously) seeing it working, before submitting the PR. Maybe I was tired or something and made a last minute change, but I should have immediately caught this.

And I'm also surprised not to see any complaints that scroll-to-top stopped working (whereas there _are_ complaints that smoothly expanding is annoying).

[the remark this answers is not in the document]
Comment (dterrahe) [reply src/dtgtk/expander.c:133]: Seems I "simplified" from 
`  const int move = MAX(scroll_to_top ? from_top : from_top - MAX(0, MIN(from_top, spare)),
                       - MAX(0, spare - from_top));
`
at the last minute, without realising that's not the same thing (and doing sloppy testing). Apologies.

But now I also identified two more corner cases:
- when expanding a module below the currently open (and therefore now collapsing) module, I'm trying to keep it in the same place. But afterwards, the user can move it. If it then changes size, it again will be brought back to where it was when it was expanded. This should not happen.
- both left (lib) and right (iop) modules can be moving at the same time (for example because the history size changed). This can cause modules to move (upward) when they should be staying in the same place.

I'm working on solutions for all three.
